learning/tour-of-beam/frontend/lib/repositories/user_progress/abstract.dart

```diff
@@ -16,30 +16,28 @@
  * limitations under the License.
  */
 
-import 'package:flutter/widgets.dart';
-import 'package:get_it/get_it.dart';
+import 'package:playground_components/playground_components.dart';
 
-import '../../../cache/unit_progress.dart';
-import '../../../repositories/client/client.dart';
+import '../models/get_user_progress_response.dart';
 
-class UnitController extends ChangeNotifier {
-  final String unitId;
-  final String sdkId;
+abstract class AbstractUserProgressRepository {
+  Future<GetUserProgressResponse?> getUserProgress(
+    String sdkId,
+  );
 
-  UnitController({
-    required this.unitId,
-    required this.sdkId,
+  Future<void> completeUnit(
+    String sdkId,
+    String unitId,
+  );
+
+  Future<void> saveUnitSnippet({
+    required Sdk sdk,
+    required List<SnippetFile> snippetFiles,
+    required String unitId,
   });
 
-  Future<void> completeUnit() async {
-    final client = GetIt.instance.get<TobClient>();
-    final unitProgressCache = GetIt.instance.get<UnitProgressCache>();
-    try {
-      unitProgressCache.addUpdatingUnitId(unitId);
-      await client.postUnitComplete(sdkId, unitId);
-    } finally {
-      await unitProgressCache.updateUnitProgress();
-      unitProgressCache.clearUpdatingUnitId(unitId);
-    }
-  }
+  Future<ExampleLoadingDescriptor> getSavedSnippet({
```

Comment:
`getSavedSnippet` only gets the saved snippet, while `getUnitSnippet` would infer that it can get any type of snippet. `getSavedUnitSnippet` might be understood as "get the snippet of a saved unit".

Reply:
`getSavedSnippetDescriptor` then.

Reply:
We don't refer to SnippetDescriptor elsewhere. I don't think `getSavedSnippetDescriptor` is better than `getSavedSnippet`. Maybe to rename `saveUnitSnippet` to `saveSnippet`?

Reply:
We don't refer to SnippetDescriptor elsewhere. I don't think `getSavedSnippetDescriptor` is better than `getSavedSnippet`. Maybe to rename `saveUnitSnippet` to `saveSnippet`?

Reply:
`getSavedExampleLoadingDescriptor`